### PR TITLE
Fix element binding issue for multiple custom node instances

### PR DESCRIPTION
**Pull request by ke-yu**
### Purpose

This PR is to fix defect MAGN-10165 that element binding doesn't work for custom node.

Element binding is to associate elements with node that *directly* generates those elements. In other word, the type of node should be `DSFunction` (function from zero touch library). If node generates some elements, when .dyn file is saved, there is a `<SessionTraceData>` which records node ID and encoded elements' information:

```xml
<SessionTraceData>
    <NodeTraceData NodeId="ccb31126-9527-45be-8ac5-ba34e46ea436">
        <CallsiteTraceData>....</CallsiteTraceData>
    </NodeTraceData>
</SessionTraceData>
```

When this .dyn is re-opened, trace data will be loaded. At runtime, when executing a node (function), instead of creating a new call site each time, same call site will be reused if this node has been executed before and associate the call site with loaded or updated trace data. The runtime maintains two dictionaries:  `CallSiteCache` dictionary maps from node's GUID combined with context information like function scope and class scope to call site instance and `CallSiteToNodeMap` maps call site to node. The latter will be used to get the associated trace data of a node.

Now, if there are two nodes which are custom node instances of a same custom node, and node `C` in this custom node creates elements:
```

    +---+
    | A | -------->  +---Custom Node X--+
    +---+            |                  |
                     |       +---+      |
    +---+            |       | C |      |
    | B | -------->  |       +---+      |    
    +---+            |                  | 
                     +------------------+
```

As we mentioned before, trace data is stored in call site instance. In this case, `A` eventually calls `C`, so the key value pair in `CallSiteCache` would be `{C_GUID + context information, call site instance for C}`, the key value pair in `CallSiteToNodeMap` would be `{call site instance for C, C_GUID }`. When `B` eventually calls `C`, same call site will be reused. So here there are two problems:
  1. Trace data stored in the call site is overwritten by the other call, because call site is re-used.
  2. Trace data is not associated with top level node like `A` and `B', but with node `C` which cannot be serialized.

For the first problem, instead of just using `C`'s GUID, we generate a long call site identifier (used in `CallSiteCache`) encodes the information of the whole function call chain (`A` -> `C` and `B` -> `C`) to ensure the uniqueness. For example, in this case the call site identifier may looks like:
```
    GUID_A + A's context information + GUID_C + C's context information
```

For the second problem, instead of saving trace data to the node that initiates function call, we use top level nodes -- in this case they are `A` and `B` -- and associate call sites with them. When .dyn file is saved, a node may have multiple `CallsiteTraceData`. This PR added an attribute `CallSiteID` to differentiate them:
```xml
<SessionTraceData>
    <NodeTraceData NodeId="GUID_A">
      <CallsiteTraceData CallSiteID="GUID_A_GUID_C">...</CallsiteTraceData>
      <CallsiteTraceData CallSiteID="GUID_A_GUID_D">...</CallsiteTraceData>
    </NodeTraceData>
   <NodeTraceData NodeId="GUID_B">
      <CallsiteTraceData CallSiteID="GUID_B_GUID_C">...</CallsiteTraceData>
    </NodeTraceData>
</SessionTraceData>
```

This solution has some downsides:
  1. Backward compatibility. Old .dyn file doesn't have `CallSiteID` attribute, in this case we use empty string as call site ID. If fails to look up call site instance with some non-empty call site ID, we'll try to use empty string to search.
  2. The size of .dyn file may be bloated, as a custom node instance node will aggregate all elements that generated internally.
  3. Trace data is stored in a flatten list, so searching `CallSiteCache` is not efficient, we probably need a more structural way to represent trace data (at runtime or in .dyn file).

Element binding test case run without regression:
untitled

Related Revit regression test case: https://github.com/DynamoDS/DynamoRevit/pull/1223

### Declarations

Check these if you believe they are true

- [ ] The code base is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.

### Reviewers
@sharadkjaiswal

### CODE REVIEW

**Comment by mjkkirschner** [comment]
Does this work if the custom node serializes multi ids or multiple single ids ?

**Comment by ke-yu** [comment]
@mjkkirschner I think so, multiple elements eventually will saved to a single trace data (represented in nested data).

File: src/DynamoCore/Graph/Nodes/NodeCategories.cs [hunk 1]
```diff
@@ -280,7 +281,7 @@ internal static string GetDocumentXmlPath(XmlDocument document)
         /// to be saved to the XmlDocument. This parameter cannot be null and 
         /// must represent a non-empty list of node-data-list pairs.</param>
         internal static void SaveTraceDataToXmlDocument(XmlDocument document,
-            IEnumerable<KeyValuePair<Guid, List<string>>> nodeTraceDataList)
+            IEnumerable<KeyValuePair<Guid, List<CallSite.RawTraceData>>> nodeTraceDataList)
```
Comment (sharadkjaiswal) [inline src/DynamoCore/Graph/Nodes/NodeCategories.cs]: Are these methods used by DynamoRevit? If yes then we might be breaking binary compatibility.

File: src/Engine/ProtoCore/RuntimeData.cs [hunk 2]
```diff
@@ -55,25 +48,39 @@ public RuntimeData()
...
         /// <returns></returns>
-        public CallSite GetCallSite(GraphNode graphNode,
-                                    int classScope,
-                                    string methodName,
-                                    Executable executable,
-                                    int runningBlock,
-                                    Options options,
-                                    RuntimeStatus runtimeStatus
-             )
+        public CallSite GetCallSite(int classScope, string methodName, Executable executable, RuntimeCore runtimeCore)
```
Comment (sharadkjaiswal) [inline src/Engine/ProtoCore/RuntimeData.cs]: Where is this method called from? If DynamoRevit calls it, then we are breaking the compatibility.

File: src/Engine/ProtoCore/DSASM/Executive.cs [hunk 3]
```diff
@@ -400,12 +400,12 @@ private void RestoreExecutive(int exeblock)
             pc = (int)rmem.GetAtRelative(StackFrame.FrameIndexReturnAddress).IntegerValue;
         }
 
-        private void PushInterpreterProps(InterpreterProperties properties)
+        public void PushInterpreterProps(InterpreterProperties properties)
```
Comment (sharadkjaiswal) [inline src/Engine/ProtoCore/DSASM/Executive.cs]: Why do we need to make it public?

File: src/DynamoCore/Graph/Nodes/NodeCategories.cs (hunk 1 above)
Comment (ke-yu) [inline src/DynamoCore/Graph/Nodes/NodeCategories.cs]: No. :-) It is internally used by `WorkspaceModel.SerializeSessionData`.

File: src/Engine/ProtoCore/RuntimeData.cs (hunk 2 above)
Comment (ke-yu) [inline src/Engine/ProtoCore/RuntimeData.cs]: Both from the VM runtime - one is from instruction CALLR's handler, the other one is builtin function `DotMethod()`.

File: src/Engine/ProtoCore/DSASM/Executive.cs (hunk 3 above)
Comment (ke-yu) [inline src/Engine/ProtoCore/DSASM/Executive.cs]: Long story short.

Normally before executing some code / calling a function, the VM internally will call `PushInterpreterProps` to save current executing graph node together with other information, and when returning from a function (executing instruction `RETURN`) or finishing executing code (executing instruction `RETB`), the VM will call `PopInterpreterPros` to pop context information.

But in live execution, the execution is split into two phases: executing new code and applying update. For the later no `RETB` instruction will be executed, we have to manually do cleanup. 

This change has nothing to do with this defect fixing, but just for clean-up purpose.

**Pull request merged**